src/rpc/blockchain.cpp

```diff
@@ -283,7 +284,22 @@ static RPCHelpMan waitfornewblock()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     Mining& miner = EnsureMining(node);
 
+   // Use tip as reference block, unless a block hash was provided.
     auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (!request.params[1].isNull()) {
+        uint256 hash(ParseHashV(request.params[1], "blockhash"));
+        {
+            ChainstateManager& chainman = EnsureAnyChainman(request.context);
+            LOCK(cs_main);
+            const CBlockIndex* block_index{chainman.m_blockman.LookupBlockIndex(hash)};
+            if (!block_index) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+            }
+            block.hash = block_index->GetBlockHash();
+            block.height = block_index->nHeight;
+        }
```

Comment:
Why do all this? Can't we just compare `hash` to `block` and return?

Reply:
@luke-jr it should be simpler now